### PR TITLE
Facebook Audience Network adapter

### PR DESCRIPTION
**Pull request by johngeorgewright**
Just as I finished making this pull request I realised that there's an `ajax.js` file. I've used a small library called [reqwest](https://github.com/ded/reqwest) which covers JSONP and CORs (which is what's required here). Would you prefer using a library or keeping something custom?

### CODE REVIEW

**Comment by mercuryyy** [comment]
Hey John, this adapter looks great, is there any way you can add it to http://prebid.org/download.html
i'm not really sure how to build from source using rubicon,aol and this additional adapter

**Comment by protonate** [comment]
The principle concern with using libraries in the production build is file size of prebid.js distributable. How much weight does reqwest add to the build?

@mercuryyy, adapters are made available on the download page following a version release, and we should be releasing v0.12.0 in two to three weeks. You can build Prebid yourself by taking the steps detailed [here](https://github.com/prebid/Prebid.js#build).

**Comment by johngeorgewright** [comment]
Hey @protonate, I'm sure we can make something smaller than a full ajax library. So I'll add to the ajax file you already have for CORs options tomorrow when I'm back at work.

**Comment by mercuryyy** [comment]
Cant wait to try this hope to see it in the next distribution.

**Comment by mercuryyy** [comment]
@johngeorgewright just a thought here, if you havent already implemented something simular maybe to save on latency instead of loading the whole facebook adapter request shoot a pixel to user check if he is logged in to facebook using facebook api if true then exec.

since facebook will only serv ads to logged in facebook users 95% of the time.

**Comment by johngeorgewright** [comment]
I've just pushed another version without using the 3rd party library.

I'm also in contact with FaceBook to make sure I've covered everything they're expecting... I'll let you know once I get sign off.

**Comment by mercuryyy** [comment]
I'm having problems with node.js  and compiling a pull. can you please make a pull for a version with this and rubicon project pull so i can test this out.

File: src/adapters/fan.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+const bidFactory = require('../bidfactory');
+const bidManager = require('../bidmanager');
+const BASE_URL = 'https://an.facebook.com/v1/placementbid.json';
+const ACCEPTED_FORMATS = ['320x50', '300x250', '728x90'];
+
+function sizesToFormats(sizes) {
+  return sizes
+    .map(([w, h]) => {
+      let format = `${w}x${h}`;
+      return ~ACCEPTED_FORMATS.indexOf(format) ? format : 'native';
```
Comment (protonate) [inline src/adapters/fan.js]: Is it possible to use `Array.prototype.includes` here? The IE polyfill fix in #464 should merge ahead of this PR.

**Comment by protonate** [comment]
Thanks @johngeorgewright this PR looks good to me. I've tested functionality with a setup to back FAN with DFP, and I'm seeing the DFP creatives rendered to the test page. This just confirms the ends of the setup and not if FAN itself is able to fill. Would you provide some more info on the config scenario for FAN you intend?

**Comment by johngeorgewright** [comment]
Hey @protonate, where's the normal place to note down adapter parameters and examples?

**Comment by mkendall07** [comment]
@johngeorgewright 

You can add the example to this file: https://github.com/prebid/Prebid.js/blob/master/integrationExamples/gpt/pbjs_partial_refresh_gpt.html#L129

**Comment by protonate** [comment]
Thanks for the `includes` change, the Travis-CI test failure should be resolved by rebasing your fork branch onto current master.

**Comment by johngeorgewright** [comment]
Looks like pull #464 hasn't been merged in yet... hence the fails.

**Comment by protonate** [comment]
Ok #464 is merged, please rebase again and we should be green.

**Comment by johngeorgewright** [comment]
I've been informed that it's required to send Facebook's `bid_id` as a parameter. I tried setting it to the bid object's `.ad` property, but I don't see it being added to the DFP requests.

What's my best approach? Adding some bidder settings for fan only?

**Comment by mkendall07** [comment]
@johngeorgewright 
There is a bid adapter API for setting keys via adapter. Something like this:
```var setting = { adserverTargeting: [
    {
      key: 'bid_id',
      val: function (bidResponse) {
        return bid.bid_id;
      }
    }
  ]
};
bidmanager.registerDefaultBidderSetting(bidderCode, setting);```

Note - you still need to set `bid_id` on the bid object to reference it in this API.

Why is `bid_id` always required though?

**Comment by johngeorgewright** [comment]
Thanks @mkendall07 I has actually just stumbled across that.

Not 100% sure why it's needed as I'm not hearing this directly from Facebook, but proxied through my commercial team. I have asked for a little more info on that and I'll get back to you.

**Comment by johngeorgewright** [comment]
From Facebook:

> The reason we need that is to get the ad that we cache it when the initial bid response was made, we use that bid_id to then go and fetch that ad.

**Comment by johngeorgewright** [comment]
Ah, I now understand that the bid_id only required to build the ad. We can do that inside Prebid.js instead.

**Comment by johngeorgewright** [comment]
After a bit of a chat with Facebook, I realised that I was supposed to include the ad templates (doh). I've just pushed a change to build the ad templates. I've seen it working on a demo site 😄

File: src/adapters/fan.js [hunk 2]
```diff
@@ -1,17 +1,139 @@
...
+    if (ACCEPTED_FORMATS.includes(size)) {
+      return size;
+    }
+  }
+  return 'native';
+}
+
+function buildAd(unit, bid) {
+  return getPlacementWebAdFormat(unit) === 'native' ?
+    `<head>
```
Comment (mkendall07) [inline src/adapters/fan.js]: Did you build these templates or are they from FB?

**Comment by mercuryyy** [comment]
},{
    bidder: 'fan',
    params: {
    placementId: '1679678852310121_1751482161796456'
   }
 }]
},{

MESSAGE: CALLING BIDDER ======= fan

but no bids are ever called (im logged in to facebook direct tags are showing if i test them) am i missing anything for those saying they see ads rendering.
Do i need to enable anything on facebook side? contact facebook ?

File: src/adapters/fan.js (hunk 2 above)
Comment (johngeorgewright) [inline src/adapters/fan.js]: FB gave me the templates.

**Comment by johngeorgewright** [comment]
@mercuryyy As long as that's a correct Audience Network placement ID then it should work. 2 things to try:

1. Look at your debugger's network tab and search for `an.facebook`. You should see an Ajax request made with a JSON response. Although not always helpful, the JSON response may contain an error message.
1. Add the `testMode` parameter to the params:
```javascript
{
  bidder: 'fan',
  sizes: [[300, 250]],
  params: {
    placementId: '1679678852310121_1751482161796456',
    testMode: true
  }
}
```

**Comment by mercuryyy** [comment]
@johngeorgewright i tried not getting any an.facebook ajax request. 

I looked into the compressed built and i do see all the facebok sdk's and adapters added on to the build so that part is correct.

MESSAGE: Emitting event for: auctionInit
pbf.js:1 MESSAGE: CALLING BIDDER ======= rubicon
pbf.js:1 MESSAGE: Emitting event for: bidRequested
pbf.js:1 MESSAGE: CALLING BIDDER ======= fan
pbf.js:1 MESSAGE: Emitting event for: bidRequested
pbf.js:1 MESSAGE: Emitting event for: bidAdjustment
pbf.js:1 MESSAGE: Emitting event for: bidResponse
pbf.js:1 INFO: Invoking pbjs.setTargetingForGPTAsync 
pbf.js:1 MESSAGE: Rubicon Project bidding complete: 2845
pbf.js:1 MESSAGE: Emitting event for: bidAdjustment
pbf.js:1 MESSAGE: Emitting event for: bidResponse
pbf.js:1 MESSAGE: Emitting event for: bidAdjustment
pbf.js:1 MESSAGE: Emitting event for: bidResponse
pbf.js:1 MESSAGE: Emitting event for: bidAdjustment
pbf.js:1 MESSAGE: Emitting event for: bidResponse
pbf.js:1 MESSAGE: Prebid auction cleared
pbf.js:1 MESSAGE: Emitting event for: auctionEnd

**Comment by johngeorgewright** [comment]
@mercuryyy I think I can see a potential bug. If you don't specify a size for your placement then that might cause an issue. Could that be the problem?

**Comment by mercuryyy** [comment]
Size is specified, please see - http://ow.ly/TwEw303299H

**Comment by mercuryyy** [comment]
ok i found it

{ "errors": [ "Application not authorised for header bidding" ], "bids": { } }

**Comment by johngeorgewright** [comment]
Just about to send you a screenshot... yep, that's your problem.

**Comment by mercuryyy** [comment]
so we do need permission from facebook ?

**Comment by johngeorgewright** [comment]
Yes. You need an active account/application.

**Comment by mercuryyy** [comment]
I do have an active account/application if tags run live they run fine. am i missing something ?

**Comment by mercuryyy** [comment]
https://cloud.githubusercontent.com/assets/7463132/17484641/e0fdc180-5d58-11e6-9a77-372f6c879340.png

**Comment by johngeorgewright** [comment]
Sorry @mercuryyy, I can't help much further with that. You may need to raise it with FaceBook themselves. I have a feeling there was a small hiccup when we first started the application too.

**Comment by mercuryyy** [comment]
Yeah i guess they need to manually enabled header bidding on some accounts thanks John. what would be the best contact i saw you mentioned you where chatting with them?

**Comment by mercuryyy** [comment]
Thank john, contacted facebook. will update when i get word.

File: integrationExamples/gpt/pbjs_partial_refresh_gpt.html [hunk 3]
```diff
@@ -82,6 +82,11 @@
                             publisherId: 'TO ADD',
                             adSlot: 'TO ADD@300x600'
                         }
+                    }, {
+                        bidder: 'fan',
+                        params: {
+                            placementId: 'TO ADD'
```
Comment (protonate) [inline integrationExamples/gpt/pbjs_partial_refresh_gpt.html]: should `testMode: true` be added here as option for development?

Comment (johngeorgewright) [inline integrationExamples/gpt/pbjs_partial_refresh_gpt.html]: That would mean that Facebook would always return with an ad at a very high CPM... so yes, if that would help with the demo.

**Comment by mercuryyy** [comment]
@johngeorgewright facebook got back to me saying they have no idea what im talking about, they dont know about header bidding or the https://an.facebook.com/v1/placementbid.json  file ;/ 
what am i missing here? whats the correct way to approach the right department/person ?

**Comment by mercuryyy** [comment]
@johngeorgewright hi facebook is saying its invite only, can you shoot an email to your contact at facebook to consider including me 'd really appreciate it. APP ID: 1679678852310121 (account hardly has impression just signup now but 50m+ monthly impression on the header) many many thanks.

**Comment by mkendall07** [comment]
@mercuryyy Please take your comments offline, and don't share sensitive data in public threads.

**Comment by mercuryyy** [comment]
@johngeorgewright i tried contacting facebook support in various deferent ways, they have very little to none information about this and say its limited to beta accounts. can you prehpas direct me to the correct person/department direct contact information, would be really appreciated.

**Pull request merged**